### PR TITLE
Feature/299847 date of visit

### PR DESCRIPTION
**Pull request by aldeez988**

### CODE REVIEW

File: app/routes/endemics/date-of-visit.js [hunk 1]
```diff
@@ -168,22 +168,16 @@ module.exports = [
...
-
-        if (lastReviewDate) {
-          const tenMonthsSinceLastReview = new Date(lastReviewDate)
-          tenMonthsSinceLastReview.setMonth(tenMonthsSinceLastReview.getMonth() + config.endemicsClaimExpiryTimeMonths)
-
-          if (dateOfVisit < tenMonthsSinceLastReview) {
-            return h.view(endemicsDateOfVisitException, { backLink: pageUrl, ruralPaymentsAgency: config.ruralPaymentsAgency }).code(400).takeover()
-          }
+        const { isValid, content } = typeOfReview === claimType.review ? isValidReviewDate(previousClaims, dateOfVisit) : isValidEndemicsDate(previousClaims, dateOfVisit)
+        console.log('****************', previousClaims, dateOfVisit)
```
Comment (alex742) [inline app/routes/endemics/date-of-visit.js:178]: Remove console.log

File: app/api-requests/claim-service-api.js [hunk 2]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
   }
 }
 
+function is10MonthsDifference (firstDate, secondDate, comparisonOperator = '') {
+  const tenMonthsSinceLastClaim = new Date(secondDate)
+  tenMonthsSinceLastClaim.setMonth(tenMonthsSinceLastClaim.getMonth() + config.endemicsClaimExpiryTimeMonths)
+  console.log('is10MonthsDifference', firstDate, tenMonthsSinceLastClaim)
```
Comment (alex742) [inline app/api-requests/claim-service-api.js:70]: Remove console.log

File: app/api-requests/claim-service-api.js [hunk 3]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
...
+  const nextReviewClaims = (previousClaims ?? []).filter((previousClaim) => statusesFor10MonthCheck.includes(previousClaim.statusId) && previousClaim.type === claimType.review && new Date(previousClaim.data.dateOfVisit) > dateOfVisit)
+  const sortedPriorReviewClaims = _.sortBy(priorReviewClaims, [(priorReviewClaim) => { return new Date(priorReviewClaim.data.dateOfVisit) }])
+  const sortedNextReviewClaims = _.sortBy(nextReviewClaims, [(nextReviewClaim) => { return new Date(nextReviewClaim.data.dateOfVisit) }])
+  const IsValidPriorReviewDifference = sortedPriorReviewClaims.length ? is10MonthsDifference(dateOfVisit, sortedPriorReviewClaims[sortedPriorReviewClaims.length - 1].data.dateOfVisit) : undefined
+  const IsValidNextReviewDifference = sortedNextReviewClaims.length ? is10MonthsDifference(new Date(sortedNextReviewClaims[sortedNextReviewClaims.length - 1].data.dateOfVisit), dateOfVisit) : undefined
+
+  console.log('$$$$$$$$$', sortedNextReviewClaims, priorReviewClaims)
```
Comment (alex742) [inline app/api-requests/claim-service-api.js:85]: Remove console.log

File: app/api-requests/claim-service-api.js [hunk 4]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
...
+  const sortedNextEndemicsClaims = _.sortBy(nextEndemicsClaims, [(nextEndemicsClaim) => { return new Date(nextEndemicsClaim.data.dateOfVisit) }])
+  const isValidPriorFailedReviewClaimsDifference = sortedPriorFailedReviewClaims.length ? is10MonthsDifference(dateOfVisit, sortedPriorFailedReviewClaims[sortedPriorFailedReviewClaims.length - 1].data.dateOfVisit, 'lessThanTenMonths') : undefined
+  const isValidPriorSuccessFulReviewClaimsDifference = sortedPriorSuccessFulReviewClaims.length ? is10MonthsDifference(dateOfVisit, sortedPriorSuccessFulReviewClaims[sortedPriorSuccessFulReviewClaims.length - 1].data.dateOfVisit, 'lessThanTenMonths') : undefined
+  const isValidPriorEndemicsDifference = sortedPriorEndemicsClaims.length ? is10MonthsDifference(dateOfVisit, sortedPriorEndemicsClaims[sortedPriorEndemicsClaims.length - 1].data.dateOfVisit) : undefined
+  const isValidNextEndemicsDifference = sortedNextEndemicsClaims.length ? is10MonthsDifference(new Date(sortedNextEndemicsClaims[sortedNextEndemicsClaims.length - 1].data.dateOfVisit), dateOfVisit) : undefined
+
+  console.log('%%%%%%%%%%%%%%%', sortedPriorSuccessFulReviewClaims.length, priorSuccessFulReviewClaims)
```
Comment (alex742) [inline app/api-requests/claim-service-api.js:107]: Remove console.log

File: app/api-requests/claim-service-api.js [hunk 5]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
...
+  if (sortedPriorFailedReviewClaims.length && !isValidPriorFailedReviewClaimsDifference) {
+    return { isValid: false, content: { url: '', text: 'The Dairy Farm - SBI 123456789 had a failed review claim for beef cattle in the last 10 months.' } }
```
Comment (alex742) [inline app/api-requests/claim-service-api.js:109]: This is not static content - we are meant to be displaying the business name, SBI, and type of species that they are reviewing for

File: app/api-requests/claim-service-api.js [hunk 6]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
...
+  }
+
+  if (sortedPriorSuccessFulReviewClaims.length && !isValidPriorSuccessFulReviewClaimsDifference) {
+    return { isValid: false, content: { url: 'https://fcp-ahwr-prototype.herokuapp.com/v25/farmer/guidance-claim', text: 'There must be no more than 10 months between your annual health and welfare reviews and endemic disease follow-ups.' } }
```
Comment (alex742) [inline app/api-requests/claim-service-api.js:113]: This is a link to a prototype content page - is this a guidance page we have?

File: app/api-requests/claim-service-api.js [hunk 7]
```diff
@@ -63,9 +64,69 @@ function getMostRecentReviewDate (previousClaims, latestVetVisitApplication) {
...
+  }
+
+  if ((sortedPriorEndemicsClaims.length && !isValidPriorEndemicsDifference) ||
+    (sortedNextEndemicsClaims.length && !isValidNextEndemicsDifference)) {
+    return { isValid: false, content: { url: 'https://fcp-ahwr-prototype.herokuapp.com/v25/farmer/guidance-claim', text: 'There must be at least 10 months between your endemics follow-ups.' } }
```
[the remark this answers is not in the document]
Comment (aldeez988) [reply app/api-requests/claim-service-api.js:115]: removed

File: app/api-requests/claim-service-api.js (hunk 4 above)
Comment (aldeez988) [reply app/api-requests/claim-service-api.js]: removed

File: app/api-requests/claim-service-api.js (hunk 3 above)
Comment (aldeez988) [reply app/api-requests/claim-service-api.js]: removed

File: app/api-requests/claim-service-api.js (hunk 2 above)
Comment (aldeez988) [reply app/api-requests/claim-service-api.js]: removed

File: app/routes/endemics/date-of-visit.js (hunk 1 above)
Comment (aldeez988) [reply app/routes/endemics/date-of-visit.js]: removed

**Review by alex742 — APPROVED** [review]

**Pull request merged**